PythonVisualizations/Heap.py

```diff
@@ -0,0 +1,284 @@
+        arrayCoords = add_vector(cell_coords,
+                                 (half_border-1, half_border-1,
+                                  self.CELL_BORDER - half_border-1, self.CELL_BORDER - half_border))
+        rect = self.canvas.create_rectangle(arrayCoords,
+                                            fill=None, outline=self.CELL_BORDER_COLOR, width=self.CELL_BORDER, tags= "arrayCell")
+        self.canvas.lower(rect)
+
+        return rect
+
+    def createCellValue(self, indexOrCoords, key, color=None):
+        """Create new canvas items to represent a cell value.  A square
+        is created filled with a particular color with an text key centered
+        inside.  The position of the cell can either be an integer index in
+        the Array or the bounding box coordinates of the square.  If color
+        is not supplied, the next color in the palette is used.
+        An event handler is set up to update the VisualizationApp argument
+        with the cell's value if clicked with any button.
+        Returns the tuple, (square, text), of canvas items
+        """
+        # Determine position and color of cell
+        if isinstance(indexOrCoords, int):
+            rectPos = self.cellCoords(indexOrCoords)
+            valPos = self.cellCenter(indexOrCoords)
+        else:
+            rectPos = indexOrCoords
+            valPos = divide_vector(add_vector(rectPos[:2], rectPos[2:]), 2)
+
+        if color is None:
+            # Take the next color from the palette
+            color = drawable.palette[Heap.nextColor]
+            Heap.nextColor = (Heap.nextColor + 1) % len(drawable.palette)
+        cell_rect = self.canvas.create_rectangle(
+            *rectPos, fill=color, outline='', width=0)
+        cell_val = self.canvas.create_text(
+            *valPos, text=str(key), font=self.SMALL_FONT, fill=self.VALUE_COLOR)
+        handler = lambda e: self.setArgument(str(key))
+        for item in (cell_rect, cell_val):
+            self.canvas.tag_bind(item, '<Button>', handler)
+
+        return cell_rect, cell_val
+
+    def display(self):
+        self.canvas.delete("all")
+
+        for i in range(self.size):  # Draw grid of cells
+            self.createArrayCell(i)
+
+        # go through each Drawable in the list
+        for i, n in enumerate(self.list):
+            # create display objects for the associated Drawables
+            n.display_shape, n.display_val = self.createCellValue(
+                i, n.val, n.color)
+            n.color = self.canvas.itemconfigure(n.display_shape, 'fill')
+
+        self.window.update()
+
+    def makeButtons(self):
+        vcmd = (self.window.register(numericValidate),
+                '%d', '%i', '%P', '%s', '%S', '%v', '%V', '%W')
+
+        # numArguments decides the side where the button appears in the operations grid
+        insertButton = self.addOperation(
+            "Insert", lambda: self.clickInsert(), numArguments=1)
+        newHeapButton = self.addOperation(
```

Comment:
currently, the textbox lets you type in any character- even non-digit ones. To change this, pass in the `validationCmd = vcmd` as a part of `self.addOperation`